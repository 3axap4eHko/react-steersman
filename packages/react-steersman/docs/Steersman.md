# &lt; Steersman &gt;

Required parent component for all steersman components.

## Description

`<Steersman>` provides access to the history management and some common settings for routing.

## Props

`<Steersman>` has all props of [&lt; Route &gt;](Route.md) component and also:
#### history: object.isRequired
A history instance. Can be: browser, hash and memory 

## Example

```javascript
import React from 'react';
import { render } from 'react-dom';
import Steersman from 'react-steersman/Steersman';
import createBrowserHistory from 'react-steersman/createBrowserHistory';

const history = createBrowserHistory();

render(
  <Steersman history={history} />,
  document.getElementById('app'),
);
```