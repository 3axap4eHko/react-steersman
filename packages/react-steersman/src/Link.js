import React, { Component } from 'react';
import { object } from 'prop-types';
import Match from './Match';
import { linkPropTypes, linkDefaultProps, isReactNative } from './propTypes';

export default class Link extends Component {
  static propTypes = linkPropTypes;

  static defaultProps = linkDefaultProps;

  static contextTypes = {
    history: object.isRequired,
  };

  navigate = (event) => {
    const { history } = this.context;
    const { onPress, onClick } = this.props;
    if (history.location.pathname !== this.props.to) {
      history.push(this.props.to);
    }
    if (isReactNative) {
      onPress(event);
    } else {
      event.preventDefault();
      onClick(event);
    }
  };

  render() {
    const {
            component: Component,
            to,
            exact,
            strict,
            style,
            activeStyle,
            className,
            activeClassName,
            title,
            ...props
          } = this.props;

    const componentProps = match => ({
      match,
      ...(isReactNative
          ? {
            style: [style, match && activeStyle],
            onPress: this.navigate,
            title,
          }
          : {
            href: to,
            style: {
              ...style,
              ...(match ? activeStyle : {}),
            },
            className: [className, match && activeClassName].filter(Boolean).join(' '),
            onClick: this.navigate,
            children: title,
          }
      ),
    });

    return <Match
      path={to}
      exact={exact}
      strict={strict}
      children={({ match }) => (
        <Component
          {...props}
          {...componentProps(match)}
        />
      )}
    />;
  }
}
