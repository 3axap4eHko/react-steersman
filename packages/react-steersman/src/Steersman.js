import React, { Component } from 'react';
import { bool, object, func, oneOf } from 'prop-types';
import Transition from 'react-steersman-transition/Transition';

function DefaultTransition({ children, ...props }) {
  return (
    <Transition {...props} timeout={375}>
      {transition => children(transition)}
    </Transition>
  );
}

const nop = () => {};

export default class Steersman extends Component {

  static propTypes = {
    history: object.isRequired,
    routeTransition: func,
    onRouteUpdated: func,
    onRouteEnter: func,
    onRouteEntering: func,
    onRouteEntered: func,
    onRouteExit: func,
    onRouteExiting: func,
    onRouteExited: func,
  };

  static defaultProps = {
    routeTransition: DefaultTransition,
    onRouteUpdated: nop,
    onRouteEnter: nop,
    onRouteEntering: nop,
    onRouteEntered: nop,
    onRouteExit: nop,
    onRouteExiting: nop,
    onRouteExited: nop,
  };

  static childContextTypes = {
    history: object,
    isMounted: func,
    routeTransition: func,
    onRouteUpdated: func,
    onRouteEnter: func,
    onRouteEntering: func,
    onRouteEntered: func,
    onRouteExit: func,
    onRouteExiting: func,
    onRouteExited: func,
  };

  mounted = false;

  onRouteUpdated = (routeProps, routeState) => {
    const { onRouteUpdated } = this.props;
    onRouteUpdated(routeProps, routeState);
  };

  getChildContext() {
    return {
      history: this.props.history,
      isMounted: () => this.mounted,
      routeTransition: this.props.routeTransition,
      onRouteUpdated: this.props.onRouteUpdated,
      onRouteEnter: this.props.onRouteEnter,
      onRouteEntering: this.props.onRouteEntering,
      onRouteEntered: this.props.onRouteEntered,
      onRouteExit: this.props.onRouteExit,
      onRouteExiting: this.props.onRouteExiting,
      onRouteExited: this.props.onRouteExited,
    };
  }

  componentDidMount() {
    this.mounted = true;
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  render() {
    const { children } = this.props;
    return children;
  }
}
