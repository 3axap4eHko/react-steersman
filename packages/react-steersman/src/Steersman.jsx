import React, { Component } from 'react';
import { object, func, oneOf } from 'prop-types';

export default class Steersman extends Component {

  static propTypes = {
    history: object.isRequired,
    onLocationChanged: func,
    onRouteUpdated: func,
  };

  static defaultProps = {
    onLocationChanged: () => {},
    onRouteUpdated: () => {},
  };

  static childContextTypes = {
    history: object,
    onRouteUpdated: func,
  };

  onLocationChanged = (location, action) => {
    const { onLocationChanged } = this.props;
    if (onLocationChanged) {
      onLocationChanged(location, action);
    }
  };

  onRouteUpdated = (routePath, match, pathname) => {
    const { onRouteUpdated } = this.props;
    if (onRouteUpdated) {
      onRouteUpdated(routePath, match, pathname);
    }
  };

  getChildContext() {
    return {
      history: this.props.history,
      onRouteUpdated: this.onRouteUpdated,
    };
  }

  componentWillMount() {
    this.unlisten = this.props.history.listen(this.onLocationChanged);
  }

  componentWillUnmount() {
    this.unlisten();
  }

  render() {
    const { children } = this.props;
    return children;
  }
}
