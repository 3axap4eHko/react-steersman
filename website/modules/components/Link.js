import React from 'react';
import classNames from 'classnames';
import createLink from 'react-steersman/createLink';
import withStyles from 'react-jss';
import Text from 'react-typographic';

const Link = createLink(({ classes, className, to, title, navigate, active }) => (
  <Text
    type="link"
    href={to}
    className={classNames(classes.root, active && classes.activeRoot, className)}
    onClick={e => {
      e.preventDefault();
      navigate(to);
    }}
  >
    {title}
  </Text>
));

const styles = ({ palette }) => ({
  root: {
    color: palette.primaryTextColor,
    padding: 10,
    '&:hover': {
      color: palette.secondaryLightColor,
    },
  },
  activeRoot: {
    color: palette.secondaryLightColor,
  },
});

export default withStyles(styles)(Link);