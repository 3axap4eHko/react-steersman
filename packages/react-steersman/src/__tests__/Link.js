import React from 'react';
import renderer from 'react-test-renderer';
import createMemoryHistory from '../createMemoryHistory';
import Steersman from '../Steersman';
import Link from '../Link';

const defaultColor = '#fff';
const activeColor = '#f00';
const className = 'class';
const activeClassName = 'active';

function TestComponent(props) {
  return <a {...props} />;
}

test('Link Web active', () => {
  const context = {};
  const history = createMemoryHistory();

  context.component = renderer.create(
    <Steersman history={history}>
      <Link
        to="/"
        title="Test"
        className={className}
        activeClassName={activeClassName}
        style={{ color: defaultColor, background: '#000' }}
        activeStyle={{ color: activeColor }}
        onClick={() => {}}
        component={TestComponent}
      />
    </Steersman>,
  );
  context.tree = context.component.toTree();
  const json = context.component.toJSON();
  expect(json.props.style.color).toBe(activeColor);
  expect(json.props.className).toContain(activeClassName);
  expect(json).toMatchSnapshot();
});

test('Link Web inactive', () => {
  const context = {};
  const history = createMemoryHistory();

  context.component = renderer.create(
    <Steersman history={history}>
      <Link
        to="/test"
        title="Test"
        className={className}
        activeClassName={activeClassName}
        style={{ color: defaultColor, background: '#000' }}
        activeStyle={{ color: activeColor }}
        onClick={() => {}}
        component={TestComponent}
      />
    </Steersman>,
  );
  context.tree = context.component.toTree();
  const json = context.component.toJSON();
  expect(json.props.style.color).not.toBe(activeColor);
  expect(json.props.className).not.toContain(activeClassName);
  expect(json).toMatchSnapshot();
});
