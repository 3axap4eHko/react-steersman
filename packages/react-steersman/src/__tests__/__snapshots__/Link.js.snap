// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link Web active 1`] = `
<a
  className="class active"
  href="/"
  onClick={[Function]}
  style={
    Object {
      "background": "#000",
      "color": "#f00",
    }
  }
  title="Test"
>
  Test
</a>
`;

exports[`Link Web inactive 1`] = `
<a
  className="class"
  href="/test"
  onClick={[Function]}
  style={
    Object {
      "background": "#000",
      "color": "#fff",
    }
  }
  title="Test"
>
  Test
</a>
`;
